### PR TITLE
docs: move `not nil` to the experimental page

### PR DESCRIPTION
**Pull request by roryokane**
When I heard that this feature existed, and found the [2018 changelog entry](https://github.com/nim-lang/Nim/blob/devel/changelogs/changelog_0_19_0.md#changes-affecting-backwards-compatibility) that said `not nil` was made experimental, I looked for `not nil` documentation in https://nim-lang.org/docs/manual_experimental.html. When I didn’t find it there, I initially assumed the feature had no documentation. This change moves the documentation to where readers will expect it.

As well as moving the text to another file, I added instructions for enabling the experimental feature and tweaked some wording.

I tested this changed markup in a ReStructured Text editor and it looked mostly good, but I couldn’t confirm that the `.. code-block:: nim` will render correctly since that’s a custom block type defined by this project.

### CODE REVIEW

**Comment by roryokane** [comment]
Note to self: this change will affect the URL added to the D/Nim comparison table in https://github.com/timotheecour/D_vs_nim/pull/36/. So update that PR or make a new PR in that project after this is merged.

File: doc/manual_experimental.rst [hunk 1]
```diff
@@ -340,6 +340,36 @@ This operator will be matched against assignments to missing fields.
   a.b = c # becomes `.=`(a, b, c)
 
 
+Not nil annotation
```
Comment (timotheecour) [inline doc/manual_experimental.rst:343]: > I tested this changed markup in a ReStructured Text editor and it looked mostly good, but I couldn’t confirm that the .. code-block:: nim will render correctly since that’s a custom block type defined by this project.


`nim rst2html foo.rst` or for whole docs you can `./koch docs`

File: doc/manual_experimental.rst [hunk 2]
```diff
@@ -340,6 +340,36 @@ This operator will be matched against assignments to missing fields.
...
+
+  proc p(x: PObject) =
+    echo "not nil"
+
+  # compiler catches this:
+  p(nil)
+
+  # and also this:
+  var x: PObject
+  p(x)
```
Comment (timotheecour) [inline doc/manual_experimental.rst:367]: except it doesn't.... so I'd reword for honesty as:
```
# and also this:
=>
# but not yet this:
```

```nim
when true:
  {.experimental: "notnil"}

  type
    TObj = object
    PObject = ref TObj not nil
    TProc = (proc (x, y: int)) not nil

  proc p(x: PObject) =
    echo "not nil"
    echo x == nil

  # compiler catches this:
  # p(nil)

  # and also this:
  var x: PObject
  p(x)
```

prints:
```
not nil
true
```


maybe you could pull up some other compelling example from https://github.com/nim-lang/Nim/pull/13808 /cc @zah

File: doc/manual_experimental.rst [hunk 3]
```diff
@@ -340,6 +340,36 @@ This operator will be matched against assignments to missing fields.
...
+``{.experimental: "notnil"}``.
+
+All types for which ``nil`` is a valid value can be annotated with the ``not
+nil`` annotation to exclude ``nil`` as a valid value:
+
+.. code-block:: nim
+  {.experimental: "notnil"}
+  
+  type
+    PObject = ref TObj not nil
```
Comment (timotheecour) [inline doc/manual_experimental.rst:356]: add `TObj = object` (basically make sure it compiles )

**Comment by metagn** [comment]
Can also move [for loop macros](https://nim-lang.org/docs/manual.html#macros-for-loop-macro) and [code reordering](https://nim-lang.org/docs/manual.html#scope-rules-code-reordering).

**Comment by Araq** [comment]
This PR at least improves the situation, merging.
